### PR TITLE
Python API documentation notebook

### PR DESCRIPTION
**Pull request by KKostya**

### CODE REVIEW

File: notebooks-src/PythonAPI.md [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+- token: 8805d2e6-8fc7-49d1-a5b5-26d639755bb8
+Instance killed
+
+ Refuted, with counterexample:
+let x : int = (Z.of_nativeint (-1n))
+```
+
+When used as a context manager it initiates a pod specifically for the time within the `with` context. Once the operations within the `with` block are completed, the pod is automatically terminated and its resources are recycled.
+
+This could be limiting, especially within the Jupyter notebook environment. Alternatively one can instantiate the `imandra.session` class directly.In this case, the session remains persistent across the Jupyter cells. However, it is crucial not to forget to free the pod resources once the execution is completed by calling the `session.close()` method at the final section of the notebook.
```
Comment (actionshrimp) [inline notebooks-src/PythonAPI.md:64]: `directly.In` => `directly. In`

Might be worth rephrasing this and saying that if you don't terminate the pod, you might run into the per-user pod limit - requesting new pods will terminate old ones. Also that idle pods are terminated automatically after a timeout (the user probably isn't concerned about freeing up our resources unless it affects them).

File: notebooks-src/PythonAPI.md [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+```
+session.eval('let f x = if x > 42 then 0 else 2 * x + 1')
+```
+
+```
+EvalResponse(success=True, stdout='', stderr='')
+```
+
+Any errors (syntax, typecheking, e.t.c.) in the evaulated code will ber reported and the evaluation fails:
```
Comment (actionshrimp) [inline notebooks-src/PythonAPI.md:89]: - `evaulated => evaluated`
- `ber => be`

File: notebooks-src/PythonAPI.md [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+```
+Instance created:
+- url: https://core-europe-west1.imandra.ai/imandra-http-api/http/84d5880f-0bae-4f57-be09-101ee25a4c11
+- token: 8680adcd-6af4-4963-8f34-477736dd2568
+```
+
+## Running OCaml/ImandraML code
+
+The `eval` method of the `session` instance serves as the bridge between your Python environment and the Imandra REPL. By invoking this method, you can evaluate code within the Imandra environment.
```
Comment (actionshrimp) [inline notebooks-src/PythonAPI.md:79]: Maybe rephrase `Imandra REPL` => `Imandra session` to distinguish from `imandra-repl` itself?

File: notebooks-src/PythonAPI.md [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+    File "<user input>", line 1, characters 8-14:
+    Error: This expression has type string
+           but an expression was expected of type Z.t
+  At <user input>:1,8--14
+  1 | let x = "test" + 0
+              ^^^^^^
+```      
+
+
+The REPL environment stays persistent as long as the session remains unclosed - any declared variables or functions are staying in the context.
```
Comment (actionshrimp) [inline notebooks-src/PythonAPI.md:108]: - `The REPL environment => The Imandra session`
- `are staying in the context => remain available / stay in scope` (?)

File: notebooks-src/PythonAPI.md [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+print(session.get_history())
+```
+
+```
+## All events in session
+0. Fun: f
+1. Fun: g
+```
+
+The `session.reset()` method resets the Imandra REPL internal state, wiping all the previous variables and functions.
```
Comment (actionshrimp) [inline notebooks-src/PythonAPI.md:134]: `Imandra REPL => Imandra session`

File: notebooks-src/PythonAPI.md [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+```
+result = session.verify('fun x -> x + 1 > x')
+print(result)
+```
+
+```
+Proved
+```
+
+If the proof attempt fails, Imandra will try to synthesize a concrete counterexample illustrating the failure
```
Comment (actionshrimp) [inline notebooks-src/PythonAPI.md:160]: `failure` => `failure:`

File: notebooks-src/PythonAPI.md [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+print(result)
+```
+
+```
+Instance found:
+let x : int = (Z.of_nativeint (-1n))
+let y : int = (Z.of_nativeint (5n))
+```
+
+If the constraints are found to be unsatisfiable, the result is not
```
Comment (actionshrimp) [inline notebooks-src/PythonAPI.md:189]: `the result is not` ?

File: notebooks-src/PythonAPI.md [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+session.eval("let rec fib x = if x <= 0 then 1 else fib (x - 1)")
+result = session.instance("fun x -> x < 101 ==> fib x <> 1")
+print(result)
+```
+
+```
+Unknown: Verified up to bound 100
+```
+
+This goal is in fact a property that is better suited for verification by induction. We might try adding the `auto` hint to the above goal to invoke the Imandra's inductive waterfall and prove it.
```
Comment (actionshrimp) [inline notebooks-src/PythonAPI.md:214]: `prove it.` => `prove it:`

File: notebooks-src/PythonAPI.md [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+```
+Instance found:
+let x : int = (Z.of_nativeint (0n))
+```
+
+## Region decomposition
+
+The term Region Decomposition refers to a (geometrically inspired) "slicing" of an algorithm’s state-space into distinct regions where the behavior of the algorithm is invariant, i.e., where it behaves "the same." Each "slice" or region of the state-space describes a family of inputs and the output of the algorithm upon them, both of which may be symbolic and represent (potentially infinitely) many concrete instances.
+
+The `session.decompose(...)` method allows you to run perform the Region Decomposition given the function name:
```
Comment (actionshrimp) [inline notebooks-src/PythonAPI.md:231]: `allows you to run perform the Region Decomposition` => `allows you to perform Region Decomposition`

File: notebooks-src/PythonAPI.md [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+---------- Region 1 ----------
+Constraints
+   x <= 0
+Invariant: 
+   x
+```
+
+# Closing the Imandra session
+
+Always ensure to close the session after use.
```
Comment (actionshrimp) [inline notebooks-src/PythonAPI.md:261]: `ensure to close` => `ensure you close`

**Review by actionshrimp — CHANGES_REQUESTED** [review]

File: notebooks-src/PythonAPI.md (hunk 1 above)
Comment (KKostya) [reply notebooks-src/PythonAPI.md]: Rephrased

File: notebooks-src/PythonAPI.md (hunk 3 above)
Comment (KKostya) [reply notebooks-src/PythonAPI.md]: makes sense

File: notebooks-src/PythonAPI.md (hunk 4 above)
Comment (KKostya) [reply notebooks-src/PythonAPI.md]: Rephrased

File: notebooks-src/PythonAPI.md (hunk 7 above)
Comment (KKostya) [reply notebooks-src/PythonAPI.md]: changed.

**Review by actionshrimp — APPROVED** [review]
